### PR TITLE
Add probes cache

### PR DESCRIPTION
**Pull request by pierDipi**
This patch adds a probe cache that allows to associate a `Status`
to a given key.

The cache prunes expired entries and calls a callback when
an entry is expired, so that controllers can re-queue associated
resources.

This is part of our work to reduce e2e tests' flakiness.

## Proposed Changes

- Add probes cache

**Release Note**

```release-note
None
```

**Docs**

```
None
```

### CODE REVIEW

**Comment by aliok** [comment]
Looks good, but where's this gonna be used? Like, there's no integration right now

File: control-plane/pkg/prober/cache.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+// Status represents the resource status.
+type Status int
+
+// Cache is a key-status store.
+type Cache interface {
+	// GetStatus retries the status associated with the given key.
+	GetStatus(key string) Status
+	// UpsertStatus add or updates the status associated with the given key.
+	// Once the given key expires the onExpired callback will be called passing the arg parameter.
+	UpsertStatus(key string, status Status, arg interface{}, onExpired ExpiredFunc)
```
Comment (matzew) [inline control-plane/pkg/prober/cache.go:44]: How about something like `PutIfAbsent` ? :smile:

File: control-plane/pkg/prober/cache.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+type value struct {
+	status     Status
+	lastUpsert time.Time
+	key        string
+	arg        interface{}
+	onExpired  ExpiredFunc
+}
+
+func NewInMemoryLocalCache(ctx context.Context, expireDuration time.Duration) Cache {
```
Comment (matzew) [inline control-plane/pkg/prober/cache.go:70]: `expireDuration`  -> `expiration` ?

File: control-plane/pkg/prober/cache.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+type inMemoryLocalCache struct {
+	// mu protects targets and entries
+	mu sync.RWMutex
+	// targets is a map of key-pointer to an element in the entries list.
+	targets map[string]*list.Element
+	// entries is a doubly-linked list of all entries present in the cache.
+	// This allows fast deletion of expired entries.
+	entries *list.List
+
+	expireDuration time.Duration
```
Comment (matzew) [inline control-plane/pkg/prober/cache.go:59]: expireDuration -> expiration ?

File: control-plane/pkg/prober/cache.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+type Cache interface {
+	// GetStatus retries the status associated with the given key.
+	GetStatus(key string) Status
+	// UpsertStatus add or updates the status associated with the given key.
+	// Once the given key expires the onExpired callback will be called passing the arg parameter.
+	UpsertStatus(key string, status Status, arg interface{}, onExpired ExpiredFunc)
+}
+
+// ExpiredFunc is a callback called once an entry in the cache is expired.
+type ExpiredFunc func(key string, arg interface{})
```
Comment (matzew) [inline control-plane/pkg/prober/cache.go:48]: `ExpirationFunc`?

File: control-plane/pkg/prober/cache.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+type inMemoryLocalCache struct {
```
Comment (matzew) [inline control-plane/pkg/prober/cache.go:50]: I'd call that different, and express that the cache has some built-in `expiration`

`localExpiringCache` ? 

Do we really need the "inMemory" in the name?  If so: `inMemoryLocalExpiringCache` ?

File: control-plane/pkg/prober/cache.go (hunk 3 above)
Comment (devguyio) [reply control-plane/pkg/prober/cache.go:59]: isn't that typically called `ttl` ?

Comment (matzew) [reply control-plane/pkg/prober/cache.go:59]: not typically. I like `expiration` over `ttl`. I think expiring is _stronger_ than `ttl`

(heavily biased based on heavy usage of a sweet library like https://github.com/jhalterman/expiringmap - which this PR is bascially also trying to do)

**Comment by pierDipi** [comment]
> Looks good, but where's this gonna be used? Like, there's no integration right now

Integration in a separate PR.

This cache will be added as a reconciler field and used in `ReconcilerKind` and `FinalizeKind`.

File: control-plane/pkg/prober/cache.go (hunk 1 above)
Comment (pierDipi) [reply control-plane/pkg/prober/cache.go:44]: Should it be just `put`?
This is doing a `replace`

File: control-plane/pkg/prober/cache.go (hunk 5 above)
Comment (pierDipi) [reply control-plane/pkg/prober/cache.go]: Done, `localExpiringCache`

File: control-plane/pkg/prober/cache.go (hunk 4 above)
Comment (pierDipi) [reply control-plane/pkg/prober/cache.go]: Done!

File: control-plane/pkg/prober/cache.go (hunk 3 above)
Comment (pierDipi) [reply control-plane/pkg/prober/cache.go]: I'm using expiration, for now, I think we all get the meaning of this variable with that name.

File: control-plane/pkg/prober/cache.go (hunk 2 above)
Comment (pierDipi) [reply control-plane/pkg/prober/cache.go]: Done

File: control-plane/pkg/prober/cache.go (hunk 1 above)
Comment (pierDipi) [reply control-plane/pkg/prober/cache.go]: `PutStatus` works for me, even though I think `UpsertStatus` conveys better the idea of adding or replacing something.

**Review by matzew — APPROVED** [review]
/lgtm
/approve

thanks for addressing the comments. 

Perhaps even worth to extract as a lib :-)

**Pull request merged**